From 448b03b8630f7720a2fbd62eace61926fb373511 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <aloka.dixit@oss.qualcomm.com>
Date: Mon, 13 Jan 2025 13:14:24 -0800
Subject: [PATCH 3/4] UPSTREAM:CP:hostapd: add puncturing bitmap support in ucode

Add support to process puncturing bitmap sent by wpa_supplicant
after repeater STA completes connection. This bitmap will be
used to modify repeater AP interface configuration so that
WDS station and AP interfaces operate with same frequency
parameters.

Call flow and changes:
----------------------
apsta_state() in hostapd.uc calls freq_info() with a new
parameter for puncturing bitmap
-> uc_wpa_freq_info() creates objects for frequency parameters, add
   a new object for the puncturing bitmap.
-> uc_hostapd_iface_start() retrieves the object and overwrites the
   existing values. Update puncturing bitmap value.
   -> Configure and enable the AP interface.

Patch-dependency:
r60-02-CP-wpa_supplicant-add-puncturing-bitmap-support-in-u.patch

Signed-off-by: Aloka Dixit <aloka.dixit@oss.qualcomm.com>
---
 src/ap/ucode.c    | 8 ++++++++
 src/utils/ucode.c | 2 ++
 2 files changed, 10 insertions(+)

diff --git a/src/ap/ucode.c b/src/ap/ucode.c
index 0b27252451b7..75be8a8b757d 100644
--- a/src/ap/ucode.c
+++ b/src/ap/ucode.c
@@ -627,6 +627,12 @@ uc_hostapd_iface_start(uc_vm_t *vm, size_t nargs)
 		iface->freq = 0;
 	conf->acs = 0;
 
+	intval = ucv_int64_get(ucv_object_get(info, "punct_bitmap", NULL));
+	if (!errno)
+		conf->punct_bitmap = intval;
+	else
+		conf->punct_bitmap = 0;
+
 out:
 	switch (iface->state) {
 	case HAPD_IFACE_ENABLED:
@@ -714,6 +720,8 @@ uc_hostapd_iface_switch_channel(uc_vm_t *vm, size_t nargs)
 		csa.freq_params.center_freq1 = intval;
 	if ((intval = ucv_int64_get(ucv_object_get(info, "center_freq2", NULL))) && !errno)
 		csa.freq_params.center_freq2 = intval;
+	if ((intval = ucv_int64_get(ucv_object_get(info, "punct_bitmap", NULL))) && !errno)
+		csa.freq_params.punct_bitmap = intval;
 
 	for (i = 0; i < iface->num_bss; i++)
 		ret = hostapd_switch_channel(iface->bss[i], &csa);
diff --git a/src/utils/ucode.c b/src/utils/ucode.c
index 33c4c3abee7d..7c68c0195493 100644
--- a/src/utils/ucode.c
+++ b/src/utils/ucode.c
@@ -113,6 +113,7 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm, size_t nargs)
 	uc_value_t *chan_width = uc_fn_arg(3);
 	int cf1 = ucv_uint64_get(uc_fn_arg(4));
 	int cf2 = ucv_uint64_get(uc_fn_arg(5));
+	int punct_bitmap = ucv_uint64_get(uc_fn_arg(6));
 	int freq_val, center_idx, center_ofs;
 	enum oper_chan_width chanwidth;
 	enum hostapd_hw_mode hw_mode;
@@ -262,6 +263,7 @@ skip_chan_width:
 	ucv_object_add(ret, "center_freq2", ucv_int64_new(cf2));
 	ucv_object_add(ret, "center_seg0_idx", ucv_int64_new(seg0_idx));
 	ucv_object_add(ret, "center_seg1_idx", ucv_int64_new(seg1_idx));
+	ucv_object_add(ret, "punct_bitmap", ucv_int64_new(punct_bitmap));
 
 out:
 	return ret;
-- 
2.34.1

